### ISSUE TITLE
Issues with transfer operators

### ISSUE THREAD

**Issue by brownbaerchen** [Parallel-in-Time/pySDC#421]
I started looking into the transfer operators a little bit and noticed some issues. I will collect what I find here because I don't know if I have time to fix everything.

## Side effects of `MultiComponentMesh`
First of all, there are some side effects of the implementation of `imex_mesh` via `MultiComponentMesh`. I have seen, for instance [here](https://github.com/Parallel-in-Time/pySDC/blob/73c927442fa70f2bc2505e65fbbb0868c7301a8f/pySDC/implementations/transfer_classes/TransferMesh_FFT.py#L58):
```
if isinstance(G, mesh):
    ...
elif  isinstance(G, imex_mesh):
    ...
```
However, because an object is an instance of all classes its own class is derived from, an `imex_mesh` object is also an instance of `mesh` and the wrong condition will be entered. I expect we need to be mindful of this distinction throughout the code, not just in the transfer classes. I suggest instead:
```
if type(G).__name__ == "mesh":
    ...
elif type(G).__name__ == "imex_mesh":
    ...
```
The `type` function contains no information about inheritance and by using `__name__`, we can check against a string and avoid importing the data type. This will be handy for GPU agnostic code. Eventually, I want `if type(G).__name__ in ["mesh", "cupy_mesh"]:`.

## Want non-normalised grids
I tested the order of interpolation and restriction for some classes. Here are some limitations I found:
 - [ ] `mesh_to_mesh` interpolation and restriction work only for grids normalised to 1

Feel free to check the boxes after a respective fix. Let me know when you find more restrictions or add them to this issue yourself.

**Comment by pancetta** [comment]
Huh, interesting.. you're right, the transfer operations are not as generic as I thought.

**Comment by pancetta** [comment]
👍 for the first item.

**Comment by brownbaerchen** [comment]
> Huh, interesting.. you're right, the transfer operations are not as generic as I thought. For `mesh_to_mesh` I understand this, but are your sure about `mesh_to_mesh_fft`?

No, I think `mesh_to_mesh_fft` is fine actually. I got stuff mixed up ;)

**Comment by pancetta** [comment]
Still relevant — both boxes were open and both reproduce. Fixed in #696, along with three more things that turned up while checking the operators properly.

**Side effects of `MultiComponentMesh`:** half-fixed since you opened this. `TransferMesh`, `TransferMesh_FFT` and `TransferMesh_MPIFFT` moved to the `hasattr(type(x), 'components')` pattern at some point, but `TransferMesh_NoCoarse` and `TransferMesh_FFT2D` did not. The first returns an `imex_mesh` as a shape-`(2, N)` plain `mesh`, so the next `.impl` access raises `AttributeError`; the second fails to broadcast. Both `imex_mesh` branches were unreachable dead code.

Good news for the rest of the codebase: the `isinstance` calls left in the PETSc, FEniCS, Firedrake and particle transfer classes are all safe. `petsc_vec_imex`, `rhs_fenics_mesh`, `IMEX_firedrake_mesh`, `particles` and `fields` derive from `object`, not from the type they are tested against, so there is no inheritance to trip over. After #696 there is no `isinstance(..., mesh)` dispatch left anywhere outside `mesh.py` itself.

**Non-normalised grids:** confirmed, and the root cause is in `transfer_helper`, not in `mesh_to_mesh`. The periodic helpers hardcode a period of one (`p - np.floor(p / 1.0) * 1.0`, `cont_arr += 1`), and `mesh_to_mesh` fed them `i * dx`. Order 8 measures as 1.26 on a domain of length `2*pi`. Since the interpolation weights only depend on ratios of distances, the fix is to build the grids on the unit domain — `mesh_to_mesh` now does not read `prob.dx` at all. Side benefit: each dimension of an n-d problem is normalised by its own `nvars` instead of sharing one scalar `dx`.

You were right about `mesh_to_mesh_fft`, by the way — it is domain-length agnostic. But it had a different bug: prolongation put the coarse Nyquist coefficient at the *fine* Nyquist wavenumber. Also `TransferMesh_FFT2D` prolongation scaled by `ratio * 2` instead of `ratio**2`, so it was only ever correct at a refinement factor of two, and `fft_to_fft` sliced with a hardcoded two dimensions, so 3d raised a `ValueError`.

**On testing:** nothing anywhere checked any *property* of a transfer operator, only the interpolation order, and `test_mesh_to_mesh.py` only ever calls `prolong`. `TransferMesh_FFT2D` and `TransferMesh_NoCoarse` had no test at all. The FFT order tests use a single sine, which never excites the Nyquist mode — that is why it survived. #696 adds `test_space_transfer_properties.py` covering multi-component handling in both directions, spectral exactness up to and including Nyquist, `R = 2^-d P^T`, constant preservation, the identity without coarsening and the `rorder = 0` injection branch. Every one of them fails on the code as it is today. Also, whoever wrote `test_transfer_order.py` had already parametrized `L` and left `[1.0]` in the list for `mesh_to_mesh`; it is `[1.0, 2*pi]` now.

One gap left that I did not close: the PETSc, FEniCS and particle transfers still have no direct tests, only indirect exercise via `tutorial/step_7` and the Hamiltonian project. Their dispatch is safe but their order and adjointness are unverified. Happy to leave that for a separate issue.
